### PR TITLE
Add asserts to the gmg solver to detect nan's and inf's earlier.

### PR DESCRIPTION
**Pull request by MFraters**
This adds some asserts to find problems with wrong or poisoned values passed to the gmg solver.

### CODE REVIEW

File: source/simulator/stokes_matrix_free.cc [hunk 1]
```diff
@@ -1422,6 +1422,14 @@ namespace aspect
 
                       active_cell_data.newton_factor_wrt_pressure_table(cell,q)[i]
                         = derivatives->viscosity_derivative_wrt_pressure[q] * newton_derivative_scaling_factor;
+                      Assert(!std::isnan(active_cell_data.newton_factor_wrt_pressure_table(cell,q)[i]),
```
Comment (tjhei) [inline source/simulator/stokes_matrix_free.cc:1425]: how about using ``std::isfinite`` instead?

Comment (MFraters) [reply source/simulator/stokes_matrix_free.cc]: ah, I wasn't aware that that included nan's. Thanks! I update it.

**Review by tjhei — APPROVED** [review]

File: source/simulator/stokes_matrix_free.cc [hunk 2]
```diff
@@ -1432,6 +1436,11 @@ namespace aspect
                             active_cell_data.newton_factor_wrt_strain_rate_table(cell, q)[m][n][i]
                               = derivatives->viscosity_derivative_wrt_strain_rate[q][m][n]
                                 * newton_derivative_scaling_factor * alpha;
+
+                            Assert(std::isfinite(active_cell_data.strain_rate_table(cell, q)[m][n][i]),
+                                   ExcMessage("active_cell_data.strain_rate_table has an element which is not a finite: " + std::to_string(active_cell_data.strain_rate_table(cell, q)[m][n][i])));
```
Comment (tjhei) [inline source/simulator/stokes_matrix_free.cc:1441]:
```suggestion
                                   ExcMessage("active_cell_data.strain_rate_table has an element which is not finite: " + std::to_string(active_cell_data.strain_rate_table(cell, q)[m][n][i])));
```

File: source/simulator/stokes_matrix_free.cc [hunk 3]
```diff
@@ -1432,6 +1436,11 @@ namespace aspect
...
+                            Assert(std::isfinite(active_cell_data.newton_factor_wrt_strain_rate_table(cell, q)[m][n][i]),
+                                   ExcMessage("active_cell_data.newton_factor_wrt_strain_rate_table has an element which is not a finite: " + std::to_string(active_cell_data.newton_factor_wrt_strain_rate_table(cell, q)[m][n][i])));
```
Comment (tjhei) [inline source/simulator/stokes_matrix_free.cc:1443]:
```suggestion
                                   ExcMessage("active_cell_data.newton_factor_wrt_strain_rate_table has an element which is not finite: " + std::to_string(active_cell_data.newton_factor_wrt_strain_rate_table(cell, q)[m][n][i])));
```

File: source/simulator/stokes_matrix_free.cc [hunk 4]
```diff
@@ -1422,6 +1422,10 @@ namespace aspect
 
                       active_cell_data.newton_factor_wrt_pressure_table(cell,q)[i]
                         = derivatives->viscosity_derivative_wrt_pressure[q] * newton_derivative_scaling_factor;
+                      Assert(std::isfinite(active_cell_data.newton_factor_wrt_pressure_table(cell,q)[i]),
+                             ExcMessage("active_cell_data.newton_factor_wrt_pressure_table is not a finite: " + std::to_string(active_cell_data.newton_factor_wrt_pressure_table(cell,q)[i]) +
```
Comment (tjhei) [inline source/simulator/stokes_matrix_free.cc:1426]:
```suggestion
                             ExcMessage("active_cell_data.newton_factor_wrt_pressure_table is not finite: " + std::to_string(active_cell_data.newton_factor_wrt_pressure_table(cell,q)[i]) +
```

or ``not a finite number``

**Comment by MFraters** [comment]
Good catch. I applied the suggestions.

**Review by gassmoeller — APPROVED** [review]

**Pull request merged**